dohyun06/issue163

dohyun06 opened the pull request:
# Pull request

## 개요
<!-- 어떤 이슈를 해결하였나요? -->
mapper 삭제, dto serialization 적용

## PR Checklist

- [x] 환경에 따른 실행 여부를 확인하였나요?
- [x] 변경 내용에 관한 테스트를 진행하였나요?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 공지사항 조회 시, 응답 데이터가 더욱 직관적이고 사용자 친화적인 형태로 제공됩니다.
  * 공지사항의 제목, 본문, 첨부파일, 반응, 태그, 리마인더 등 주요 정보가 보기 쉽게 정리되어 노출됩니다.

* **리팩터링**
  * 내부 데이터 변환 및 직렬화 방식이 개선되어, 불필요한 내부 정보가 응답에 포함되지 않습니다.
  * 응답 데이터의 일관성과 가독성이 향상되었습니다.
  * 공지사항 컨트롤러에 자동 직렬화 기능이 추가되어 응답 처리 효율이 개선되었습니다.
  * 공지사항 데이터 변환 로직이 단순화되어 서비스 내 직접 DTO 생성 방식을 사용합니다.
  * 불필요해진 매퍼 클래스가 제거되어 코드 유지보수가 용이해졌습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dohyun06 committed:
fix: change mapper to dto serialization

dohyun06 committed:
fix: exclude some information when return
exclude updatedAt, lastEditedAt, deletedAt, authorId

dohyun06 committed:
fix: move async logic from dto serialization to service
transform 함수 내에서 비동기가 작동하지 않아 reaction이 정상적으로 변환되지 않음

dohyun06 committed:
fix: expose additionalContents

dohyun06 committed:
fix: add error handling

dohyun06 committed:
fix: exclude group information and refactor: change the position of dto class

dohyun06 committed:
refactor: integrate duplicate logic getting result of reaction and style: change names of dto classes

siwonpada commented on apps/api/src/notice/dto/res/generalNotice.dto.ts:
```diff
@@ -8,57 +10,153 @@ export class AuthorDto {
   views: number;
 
+  @Expose()
+  @Transform(({ obj }: { obj: GeneralNoticeDto }) =>
```
```
// address.entity.ts
import { Expose } from 'class-transformer';

export class Address {
  @Expose()
  street: string;

  @Expose()
  city: string;

  constructor(partial: Partial<Address>) {
    Object.assign(this, partial);
  }
}

// user.entity.ts
import { Expose, Type } from 'class-transformer';
import { Address } from './address.entity';

export class User {
  @Expose()
  id: number;

  @Expose()
  name: string;

  @Expose()
  @Type(() => Address) // <-- 중요: Address 객체임을 명시
  address: Address;

  constructor(partial: Partial<User>) {
    Object.assign(this, partial);
  }
}

// user.controller.ts (일부)
import { Controller, Get, UseInterceptors } from '@nestjs/common';
import { ClassSerializerInterceptor } from '@nestjs/platform-express';
import { User } from './user.entity';
import { Address } from './address.entity';

@Controller('users')
@UseInterceptors(ClassSerializerInterceptor) // <-- 컨트롤러 또는 라우트 핸들러에 인터셉터 적용
export class UserController {
  @Get(':id')
  async findOne(): Promise<User> {
    const address = new Address({ street: '123 Main St', city: 'Anytown' });
    const user = new User({ id: 1, name: 'John Doe', address: address });
    return user; // NestJS가 ClassSerializerInterceptor를 통해 직렬화합니다.
  }
}
```

위와 같은 방법이 더 낫지 않을 까요?

dohyun06 replied:
음.. 형이 올린 코드를 기준으로 address 정보를 user entity에서 처리하지말고 address entity에서 처리하는 식으로 따로 처리하는 방법을 의미하는게 맞을까요?

siwonpada replied:
네

dohyun06 replied:
crawls랑 content 비교해서 정하는 중복된 부분 따로 분리했습니다

dohyun06 replied:
더 분리할 부분이 있을까요?

siwonpada replied:
@ Transform 이 사용된 모든 부분을 수정하면 될 것 같습니다.

dohyun06 replied:
수정했습니다 확인해주세요

dohyun06 committed:
refactor: separate logics getting contents
title, langs, content, deadline

dohyun06 committed:
refactor: move the logic transforming data to service

dohyun06 committed:
refactor: move a logic transforming additionalContents to service

dohyun06 committed:
style: delete imported transform function

dohyun06 committed:
refactor: integrate super and assign functions

GanghyeonSeo commented on apps/api/src/notice/notice.service.ts:
```diff
@@ -322,4 +344,73 @@ export class NoticeService {
+    langFromDto?: string,
+    userUuid?: string,
+  ): Promise<TransformNoticeDto> {
+    const resultReaction = await firstValueFrom(
```
이 부분 rxjs 사용하지 않고 배열 내장 함수로 처리할 수 있을 것 같습니다.

dohyun06 replied:
적용했습니다

GanghyeonSeo commented on apps/api/src/notice/notice.service.ts:
```diff
@@ -57,21 +63,24 @@ export class NoticeService {
-        }),
-    );
+    ).map(async (notice) => {
+      try {
```
try catch를 사용할 필요가 있는지 잘 모르겠습니다.

dohyun06 replied:
기존 코드에서 mapper를 사용할 때 .catch()를 사용하고 있었어서 유지해놓은 것이긴 합니다.

dohyun06 replied:
없어도 될 것 같긴 합니다

dohyun06 committed:
refactor: delete error handler of serializations and change the logic not to use rxjs when getting result reactions

dohyun06 committed:
fix: fix incorrect type mapping

GanghyeonSeo commented on apps/api/src/notice/notice.service.ts:
notice.service.ts에서 해당 내용들을 고려해보면 좋을 것 같습니다.
의견 있으면 말씀해주세요..! @dohyun06 @siwonpada 

1. 사용하지 않는 함수의 import 구문 제거
2. transformExpandedNotice 함수 내에서 사용하는 reduce 함수에 타입 단언 대신 Generic 사용
3. 비동기 로직이 없는 함수에서 async, await 키워드 제거
4. 해당 파일 내에서만 사용하는 함수의 경우 private 키워드 도입

dohyun06 replied:
이견 없습니다.

dohyun06 replied:
generalNotice.dto.ts에 GeneralReactionDto가 transformNotice.dto.ts에서 밖에 안 쓰여서 해당 파일로 옮기고 export를 제거하려는데 어떨까요?

dohyun06 replied:
기존에도 export가 있었어서 함부로 지워도 될지 모르겠습니다

GanghyeonSeo replied:
위에서 말씀하신 내용 이미 옮겨진 것일까요?

dohyun06 replied:
아래 시원이 형이 말한걸 적용하다보니 transformNotice.dto.ts가 필요가 없어져서 GeneralReactionDto에 export가 제거하는 것 이외에 변경점이 없을거에요

dohyun06 replied:
더 수정할게 있을까요?

dohyun06 committed:
refactor: reflect feedback
delete useless async, await and  importing modules,
use generic method in reduce function
apply private

dohyun06 committed:
refactor: move GeneralReactionDto to transformNotice.dto.ts and add ApiProperty

siwonpada commented on apps/api/src/notice/notice.service.ts:
```diff
@@ -322,4 +319,74 @@ export class NoticeService {
+            content: mainContent.body,
+            deadline: mainContent.deadline ?? null,
+          }),
+      tags: tags.map(({ name }: { name: string }) => name),
```
여기 있는 함수의 대부분을 DTO에서 처리할 수 있을 것 같습니다. nestjs의 class serializer의 Expose관련 문서를 확인해 보세요. [문서](https://docs.nestjs.com/techniques/serialization#expose-properties)

dohyun06 replied:
@ transform으로 하지말고 getter 함수로 처리하라는 말씀인가요?

siwonpada replied:
둘 다 사용해도 상관이 없을 뿐더러, isRemineded와 imageUrl 같은 맴버변수는 getter를 통해서 DTO안에서 처리를 할 수 있을 것입니다.

siwonpada replied:
DTO내에서 처리가 가능하동작은 모두 그 안에서 처리를 하고, 진짜 정 안되는 부분이 있다면, mapper형식의 함수를 사용하는 것이 맞을 것 같습니다.

dohyun06 replied:
기존에 @ transform을 전부 수정하라 하셔서...
일단 해보겠습니다

dohyun06 committed:
refactor: restore to previous serialization code and apply getter function

dohyun06 committed:
refactor: change remindersDto to User type

dohyun06 committed:
refactor: change type of tags from object to Tag type
